daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -899,6 +899,31 @@ private[lf] object Speedy {
     }
   }
 
+  private[speedy] final case class KFoldl(
+      func: SEValue,
+      var list: FrontStack[SValue],
+      frame: Frame,
+      actuals: Actuals,
+      envSize: Int,
+  ) extends Kont
+      with SomeArrayEquals {
+    def execute(acc: SValue, machine: Machine) = {
+      list.pop match {
+        case None =>
+          machine.returnValue = acc
+        case Some((item, rest)) =>
+          machine.restoreEnv(frame, actuals, envSize)
+          // NOTE: We are "recycling" the current continuation with the
+          // remainder of the list to avoid allocating a new continuation.
+          list = rest
+          machine.pushKont(this)
+          // TODO(MH): This looks like it has some potential for further
+          // performance gains once the AST nodes related to ANF have landed.
+          machine.ctrl = SEAppAtomicFun(func, Array(SEValue(acc), SEValue(item)))
```

Comment:
@nickchapman-da Please correct me if I'm wrong: This line is the application of an `SPAP` to two values. Hence, there's no need to restore the environment for this because values don't contain references into the environment. Does that sound correct? If so, I'll remove the restoring of the environment in a fresh PR.

Reply:
Yes. I think that is correct

Reply:
What if the function allocates on the stack (because it contains a let binding)?